dev-tools/releng/src/job.rs

```diff
@@ -0,0 +1,284 @@
+    }
+
+    pub(crate) fn push_command(
+        &mut self,
+        name: impl AsRef<str>,
+        command: &mut Command,
+    ) -> Selector<'_> {
+        let name = name.as_ref().to_owned();
+        assert!(!self.map.contains_key(&name), "duplicate job name {}", name);
+        self.map.insert(
+            name.clone(),
+            Job {
+                future: Box::pin(spawn_with_output(
+                    // terrible hack to deal with the `Command` builder
+                    // returning &mut
+                    std::mem::replace(command, Command::new("false")),
+                    self.logger.clone(),
+                    self.permits.clone(),
+                    name.clone(),
+                    self.log_dir.join(&name).with_extension("log"),
+                )),
+                wait_for: Vec::new(),
+                notify: Vec::new(),
+            },
+        );
+        Selector { jobs: self, name }
+    }
+
+    pub(crate) fn select(&mut self, name: impl AsRef<str>) -> Selector<'_> {
+        Selector { jobs: self, name: name.as_ref().to_owned() }
+    }
+
+    pub(crate) async fn run_all(self) -> Result<()> {
+        self.map
+            .into_values()
+            .map(Job::run)
+            .collect::<FuturesUnordered<_>>()
+            .try_collect::<()>()
+            .await
+    }
+}
+
+impl Job {
+    async fn run(self) -> Result<()> {
+        let result: Result<(), RecvError> = self
+            .wait_for
+            .into_iter()
+            .collect::<FuturesUnordered<_>>()
+            .try_collect::<()>()
+            .await;
+        result.map_err(|_| anyhow!("dependency failed"))?;
+
+        self.future.await?;
+        for sender in self.notify {
+            sender.send(()).ok();
+        }
+        Ok(())
+    }
+}
+
+impl<'a> Selector<'a> {
+    #[track_caller]
+    pub(crate) fn after(self, other: impl AsRef<str>) -> Self {
+        let (sender, receiver) = oneshot::channel();
+        self.jobs
+            .map
+            .get_mut(&self.name)
+            .expect("invalid job name")
+            .wait_for
+            .push(receiver);
+        self.jobs
+            .map
+            .get_mut(other.as_ref())
+            .expect("invalid job name")
+            .notify
+            .push(sender);
+        self
+    }
+}
+
+macro_rules! info_or_error {
+    ($logger:expr, $result:expr, $($tt:tt)*) => {
+        if $result.is_ok() {
+            ::slog::info!($logger, $($tt)*);
+        } else {
+            ::slog::error!($logger, $($tt)*);
+        }
+    };
+}
+
+async fn run_job<F>(
+    logger: Logger,
+    permits: Arc<Semaphore>,
+    name: String,
+    future: F,
+) -> Result<()>
+where
+    F: Future<Output = Result<()>> + 'static,
+{
+    if !PERMIT_NOT_REQUIRED.contains(&name.as_str()) {
+        let _ = permits.acquire_owned().await?;
+    }
+
+    info!(logger, "[{}] running task", name);
+    let start = Instant::now();
+    let result = future.await;
```

Comment:
Surprised this doesn't spawn a new Tokio task for the future.

Reply:
I don't think it needs to? We poll it immediately here by awaiting, so I'm not sure what creating a new task gets us.

Reply:
One issue is that, while one future is being executed the others will stay suspended. This can especially be a problem if there are awaits in a loop body. https://without.boats/blog/poll-progress/

Tasks execute independently, so one of them won't be blocked on others.

Reply:
Oh, yeah, there are certainly awaits in a loop body here:

https://github.com/oxidecomputer/omicron/blob/6dbca3c52b5355b542824a20baf05eee95fa0e71/dev-tools/releng/src/hubris.rs#L44-L46

I'll put this (and I suppose `async fn reader` later in this file?) in some tasks.

Reply:
I think the readers can generally be within the same task as the jobs themselves, but yeah putting each job in a task would make sense.